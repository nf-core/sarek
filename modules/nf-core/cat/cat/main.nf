process CAT_CAT {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/pigz:2.3.4' :
        'biocontainers/pigz:2.3.4' }"

    input:
    tuple val(meta), path(files_in)

    output:
    tuple val(meta), path("${prefix}"), emit: file_out
    path "versions.yml"               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def file_list = files_in.collect { it.toString() }

    // choose appropriate concatenation tool depending on input and output format

    // | input     | output     | command1 | command2 |
    // |-----------|------------|----------|----------|
    // | gzipped   | gzipped    | cat      |          |
    // | ungzipped | ungzipped  | cat      |          |
    // | gzipped   | ungzipped  | zcat     |          |
    // | ungzipped | gzipped    | cat      | pigz     |

    // Use input file ending as default
    prefix   = task.ext.prefix ?: "${meta.id}${getFileSuffix(file_list[0])}"
    out_zip  = prefix.endsWith('.gz')
    in_zip   = file_list[0].endsWith('.gz')
    command1 = (in_zip && !out_zip) ? 'zcat' : 'cat'
    command2 = (!in_zip && out_zip) ? "| pigz -c -p $task.cpus $args2" : ''
    if(file_list.contains(prefix.trim())) {
        error "The name of the input file can't be the same as for the output prefix in the " +
        "module CAT_CAT (currently `$prefix`). Please choose a different one."
    }
    """
    $command1 \\
        $args \\
        ${file_list.join(' ')} \\
        $command2 \\
        > ${prefix}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pigz: \$( pigz --version 2>&1 | sed 's/pigz //g' )
    END_VERSIONS
    """

    stub:
    def file_list   = files_in.collect { it.toString() }
    prefix          = task.ext.prefix ?: "${meta.id}${file_list[0].substring(file_list[0].lastIndexOf('.'))}"
    if(file_list.contains(prefix.trim())) {
        error "The name of the input file can't be the same as for the output prefix in the " +
        "module CAT_CAT (currently `$prefix`). Please choose a different one."
    }
    """
    touch $prefix

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pigz: \$( pigz --version 2>&1 | sed 's/pigz //g' )
    END_VERSIONS
    """
}

// for .gz files also include the second to last extension if it is present. E.g., .fasta.gz
def getFileSuffix(filename) {
    def match = filename =~ /^.*?((\.\w{1,5})?(\.\w{1,5}\.gz$))/
    return match ? match[0][1] : filename.substring(filename.lastIndexOf('.'))
}
