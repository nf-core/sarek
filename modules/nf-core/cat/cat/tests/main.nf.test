nextflow_process {

    name "Test Process CAT_CAT"
    script "../main.nf"
    process "CAT_CAT"
    tag "modules"
    tag "modules_nfcore"
    tag "cat"
    tag "cat/cat"

    test("test_cat_name_conflict") {
        when {
            params {
                outdir   = "${outputDir}"
            }
            process {
                """
                input[0] =
                    [
                        [ id:'genome', single_end:true ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.sizes', checkIfExists: true)
                        ]
                    ]
                """
            }
        }
        then {
            assertAll(
                { assert !process.success },
                { assert process.stdout.toString().contains("The name of the input file can't be the same as for the output prefix") },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }

    test("test_cat_unzipped_unzipped") {
        when {
            params {
                outdir   = "${outputDir}"
            }
            process {
                """
                input[0] =
                    [
                        [ id:'test', single_end:true ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.sizes', checkIfExists: true)
                        ]
                    ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }


    test("test_cat_zipped_zipped") {
        when {
            params {
                outdir   = "${outputDir}"
            }
            process {
                """
                input[0] =
                    [
                        [ id:'test', single_end:true ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gff3.gz', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/alignment/last/contigs.genome.maf.gz', checkIfExists: true)
                        ]
                    ]
                """
            }
        }
        then {
            def lines = path(process.out.file_out.get(0).get(1)).linesGzip
            assertAll(
                { assert process.success },
                { assert snapshot(
                    lines[0..5],
                    lines.size(),
                    process.out.versions
                    ).match()
                }
            )
        }
    }

    test("test_cat_zipped_unzipped") {
        config './nextflow_zipped_unzipped.config'

        when {
            params {
                outdir   = "${outputDir}"
            }
            process {
                """
                input[0] =
                    [
                        [ id:'test', single_end:true ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gff3.gz', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/alignment/last/contigs.genome.maf.gz', checkIfExists: true)
                        ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("test_cat_unzipped_zipped") {
        config './nextflow_unzipped_zipped.config'
        when {
            params {
                outdir   = "${outputDir}"
            }
            process {
                """
                input[0] =
                    [
                        [ id:'test', single_end:true ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.sizes', checkIfExists: true)
                        ]
                    ]
                """
            }
        }
        then {
            def lines = path(process.out.file_out.get(0).get(1)).linesGzip
            assertAll(
                { assert process.success },
                { assert snapshot(
                    lines[0..5],
                    lines.size(),
                    process.out.versions
                    ).match()
                }
            )
        }
    }

    test("test_cat_one_file_unzipped_zipped") {
        config './nextflow_unzipped_zipped.config'
        when {
            params {
                outdir   = "${outputDir}"
            }
            process {
                """
                input[0] =
                    [
                        [ id:'test', single_end:true ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                        ]
                    ]
                """
            }
        }
        then {
            def lines = path(process.out.file_out.get(0).get(1)).linesGzip
            assertAll(
                { assert process.success },
                { assert snapshot(
                    lines[0..5],
                    lines.size(),
                    process.out.versions
                    ).match()
                }
            )
        }
    }
}
