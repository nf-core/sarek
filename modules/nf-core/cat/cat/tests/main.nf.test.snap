{
    "test_cat_unzipped_unzipped": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.fasta:md5,f44b33a0e441ad58b2d3700270e2dbe2"
                    ]
                ],
                "1": [
                    "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
                ],
                "file_out": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.fasta:md5,f44b33a0e441ad58b2d3700270e2dbe2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2023-10-16T14:32:18.500464399"
    },
    "test_cat_zipped_unzipped": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "cat.txt:md5,c439d3b60e7bc03e8802a451a0d9a5d9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
                ],
                "file_out": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "cat.txt:md5,c439d3b60e7bc03e8802a451a0d9a5d9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2023-10-16T14:32:49.642741302"
    },
    "test_cat_zipped_zipped": {
        "content": [
            [
                "MT192765.1\tGenbank\ttranscript\t259\t29667\t.\t+\t.\tID=unknown_transcript_1;geneID=orf1ab;gene_name=orf1ab",
                "MT192765.1\tGenbank\tgene\t259\t21548\t.\t+\t.\tParent=unknown_transcript_1",
                "MT192765.1\tGenbank\tCDS\t259\t13461\t.\t+\t0\tParent=unknown_transcript_1;exception=\"ribosomal slippage\";gbkey=CDS;gene=orf1ab;note=\"pp1ab;translated=by -1 ribosomal frameshift\";product=\"orf1ab polyprotein\";protein_id=QIK50426.1",
                "MT192765.1\tGenbank\tCDS\t13461\t21548\t.\t+\t0\tParent=unknown_transcript_1;exception=\"ribosomal slippage\";gbkey=CDS;gene=orf1ab;note=\"pp1ab;translated=by -1 ribosomal frameshift\";product=\"orf1ab polyprotein\";protein_id=QIK50426.1",
                "MT192765.1\tGenbank\tCDS\t21556\t25377\t.\t+\t0\tParent=unknown_transcript_1;gbkey=CDS;gene=S;note=\"structural protein\";product=\"surface glycoprotein\";protein_id=QIK50427.1",
                "MT192765.1\tGenbank\tgene\t21556\t25377\t.\t+\t.\tParent=unknown_transcript_1"
            ],
            78,
            [
                "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-22T11:51:46.802978"
    },
    "test_cat_name_conflict": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-22T11:51:29.45394"
    },
    "test_cat_one_file_unzipped_zipped": {
        "content": [
            [
                ">MT192765.1 Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/PC00101P/2020, complete genome",
                "GTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGT",
                "GTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAG",
                "TAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTTTGTCCGG",
                "GTGTGACCGAAAGGTAAGATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTTT",
                "ACAGGTTCGCGACGTGCTCGTACGTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACATCTTAAAG"
            ],
            374,
            [
                "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-22T11:52:02.774016"
    },
    "test_cat_unzipped_zipped": {
        "content": [
            [
                ">MT192765.1 Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/PC00101P/2020, complete genome",
                "GTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGT",
                "GTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAG",
                "TAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTTTGTCCGG",
                "GTGTGACCGAAAGGTAAGATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTTT",
                "ACAGGTTCGCGACGTGCTCGTACGTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACATCTTAAAG"
            ],
            375,
            [
                "versions.yml:md5,115ed6177ebcff24eb99d503fa5ef894"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-22T11:51:57.581523"
    }
}